src/profile-logic/marker-data.js

```diff
@@ -119,71 +119,192 @@ export function getSearchFilteredMarkerIndexes(
+    // checking it for negative filtering.
+    if (includeMarker && hasNegativeRegexps) {
+      includeMarker = negativeFilterMarker(
+        marker,
+        markerSchemaByName,
+        searchRegExps,
+        categoryList
+      );
+    }
+
+    if (includeMarker) {
+      // If `includeMarker` is true after both positive and negative filtering,
+      // add the marker.
+      newMarkers.push(markerIndex);
+    }
+  }
+
+  return newMarkers;
+}
+
+function positiveFilterMarker(
+  marker: Marker,
+  markerSchemaByName: MarkerSchemaByName,
+  searchRegExps: MarkerRegExps,
+  categoryList: CategoryList
+): boolean {
   // Need to assign it to a constant variable so Flow doesn't complain about
   // passing it inside a function below.
-  const regExp = searchRegExp;
+  const regExps = searchRegExps;
+
   function test(value, key) {
+    key = key.toLowerCase();
+    const fieldRegexp = regExps.fieldMap.get(key);
+    const found =
+      (regExps.generic ? regExps.generic.test(value) : false) ||
+      (fieldRegexp && fieldRegexp.positive
+        ? fieldRegexp.positive.test(value)
+        : false);
+
     // Reset regexp for each iteration. Otherwise state from previous
     // iterations can cause matches to fail if the search is global or
     // sticky.
-    regExp.lastIndex = 0;
-
-    return (
-      regExp.test(value) ||
-      (!regExp.test(key + ':') && // avoid matching the key
-        // (eg. searching for 'me' should not match all markers,
-        //  even though they all have a 'name:' field).
-        regExp.test(key + ':' + value))
+    if (regExps.generic) {
+      regExps.generic.lastIndex = 0;
+    }
+    if (fieldRegexp && fieldRegexp.positive) {
+      fieldRegexp.positive.lastIndex = 0;
+    }
+    return found;
+  }
+
+  const { data, name, category } = marker;
+
+  if (categoryList[category] !== undefined) {
+    const markerCategory = categoryList[category].name;
+    if (test(markerCategory, 'cat')) {
+      return true;
+    }
+  }
+
+  if (test(name, 'name')) {
+    return true;
+  }
+
+  if (data && typeof data === 'object') {
+    if (test(data.type, 'type')) {
+      // Check the type of the marker payload first.
+      return true;
+    }
+
+    // Now check the schema for the marker payload for searchable
+    const markerSchema = getSchemaFromMarker(
+      markerSchemaByName,
+      marker.name,
+      marker.data
     );
+    if (
+      markerSchema &&
+      markerPayloadMatchesSearch(markerSchema, marker, test)
+    ) {
+      return true;
+    }
   }
 
-  const newMarkers: MarkerIndex[] = [];
-  for (const markerIndex of markerIndexes) {
-    const marker = getMarker(markerIndex);
-    const { data, name, category } = marker;
+  return false;
+}
 
-    if (categoryList[category] !== undefined) {
-      const markerCategory = categoryList[category].name;
-      if (test(markerCategory, 'cat')) {
-        newMarkers.push(markerIndex);
-        continue;
-      }
+function negativeFilterMarker(
```

Comment:
I was wondering if we can deduce this function from the positive one. I guess the only difference is the use of `regExps.generic` in the positive one, but it is quite nested in `test`.. So probably it's not that easy (I'm sure you tried).

Reply:
Yeah, I initially wanted to have a single function that we can reuse for both positive and negative but in the end decided to split them. This is because the difference is not only the `test` function but also you have to invert all the logic below that function as well. I'll think about merging them again later but this is fine as well since the functions are not too complicated.